### ISSUE TITLE
std.Fifo.discard() makes it hard to combine BufferedReader and SeekableStream 

### ISSUE THREAD

**Issue by Opioid** [ziglang/zig#10637]
When reading files I often want to combine the benefits of BufferedReader with the feature of SeekableStream (a bit like std::ifstream in c++) so I wrote myself a wrapper for that.
One issue that can pop up when searching backwards in the file, is that you are theoretically still searching in the buffered region, but the contents have been overwritten by a previous read.
It is easy enough to work around this by invalidating the buffer (fifo.head = 0, fifo.count = 0) and letting it fill again, but it seems wasteful.
Is there a good reason for invalidating the buffer contents after reading them?

**Comment by daurnimator** [comment]
> Is there a good reason for invalidating the buffer contents after reading them?

That is the entire purpose of a FIFO, instead of an ever-growing `ArrayList`....

> theoretically still searching in the buffered region, but the contents have been overwritten by a previous read.

If you have called `discard`, then the data should be gone (and trying to read that under e.g. valgrind will have an error).

**Comment by squeek502** [comment]
Here's my solution for combining buffered reading with SeekableStream:

https://github.com/squeek502/audiometa/blob/master/src/buffered_stream_source.zig

Maybe there's a need for something like this in the standard library?

**Comment by Opioid** [comment]
> > Is there a good reason for invalidating the buffer contents after reading them?
> 
> That is the entire purpose of a FIFO, instead of an ever-growing `ArrayList`....
> 
> > theoretically still searching in the buffered region, but the contents have been overwritten by a previous read.
> 
> If you have called `discard`, then the data should be gone (and trying to read that under e.g. valgrind will have an error).

We might have a misunderstanding here. I'm not asking for a ever growing ArrayList, I'm just saying it would be nice to use the allocated memory to serve a perfectly valid use case. Why does the FIFO _have_ to overwrite data once it is read?

**Comment by squeek502** [comment]
> As far as I can tell, you have the same problem, that you can never go back in the buffer, and have to reread the file instead. Is this correct?

Yes.

> I don't see a good reason for this (unless the implementation of Fifo), but I might be missing something.

If I understand correctly, that'd need a specialized Fifo implementation that also keeps track of the 'backwards' count that's valid in the buffer, and has a way of 'seeking' forward without setting elements to undefined. Could maybe write a specialized Fifo-like specific to this use-case. Unsure if it makes sense for the 'default' Fifo implementation to have this capability or not though.

**Comment by Opioid** [comment]
> If I understand correctly, that'd need a specialized Fifo implementation that also keeps track of the 'backwards' count that's valid in the buffer, and has a way of 'seeking' forward without setting elements to undefined; could maybe write a specialized Fifo-like specific to this use-case. Unsure if it makes sense for the 'default' Fifo implementation to have this capability or not though.

I don't think so. You have all the information you need between Stream.Seekpos and Fifi.head/count. Just _not_ overwriting the fifo buffer would be enough.

> EDIT: Also unsure how much it'd really matter for the average real-world use case. In my case, I'm usually seeking backwards once/twice (from the end of the file) and then only reading forward after the seeks, which wouldn't benefit from this 'backwards buffer'. The type of seeking back/forth that'd be required to get a benefit from the backwards buffer seems like it'd be fairly unusual, but I could definitely be wrong about that.

Good point. I usually search backwards once per file, after reading 4 bytes. To make an educated guess what the file contents are and then reroute to specialized code. So I could optimize for this, which would be trivial to do.

**Comment by Opioid** [comment]
I noticed that [BufferedWriter stopped using FiFo](https://github.com/ziglang/zig/commit/7287c7482a2c694c7c7f56b9f7c1744a7ae7905f), because it is to general. Perhaps a similar argument could be made for the case of BufferedReader.
